src/sources/ldap.rs

```diff
@@ -0,0 +1,453 @@
+	fn new(config: &crate::Config) -> Result<Self> {
+		Ok(Self {
+			ldap_config: match config.sources_config.ldap_config.clone() {
+				Some(ldap) => ldap,
+				None => bail!("LDAP configuration is missing"),
+			},
+			feature_flags: config.feature_flags.clone(),
+		})
+	}
+
+	fn get_source_name(&self) -> String {
+		"LDAP".into()
+	}
+
+	async fn get_diff(&self) -> Result<SourceDiff> {
+		let cache = read_cache(&self.ldap_config.cache_path).await?;
+		let (mut ldap_client, ldap_receiver) = Ldap::new(self.ldap_config.clone().into(), cache);
+
+		let is_dry_run = self.feature_flags.is_enabled(FeatureFlag::DryRun);
+		let cache_path = self.ldap_config.cache_path.clone();
+
+		let sync_handle: tokio::task::JoinHandle<Result<_>> = tokio::spawn(async move {
+			ldap_client.sync_once(None).await.context("failed to sync/fetch data from LDAP")?;
+
+			if is_dry_run {
+				tracing::warn!("Not writing ldap cache during a dry run");
+			} else {
+				let cache = ldap_client.persist_cache().await;
+				tokio::fs::write(
+					&cache_path,
+					bincode::serialize(&cache).context("failed to serialize cache")?,
+				)
+				.await
+				.context("failed to write cache")?;
+			}
+
+			tracing::info!("Finished syncing LDAP data");
+
+			Ok(())
+		});
+
+		let (added, changed, removed) = self.get_user_changes(ldap_receiver).await?;
+
+		sync_handle.await??;
+
+		Ok(SourceDiff {
+			new_users: added,
+			changed_users: changed.into_iter().map(|(old, new)| ChangedUser { old, new }).collect(),
+			deleted_user_ids: removed,
+		})
+	}
+}
+
+impl LdapSource {
+	/// Get user changes from an ldap receiver
+	pub async fn get_user_changes(
+		&self,
+		ldap_receiver: Receiver<EntryStatus>,
+	) -> Result<(Vec<User>, Vec<(User, User)>, Vec<UserId>)> {
+		ReceiverStream::new(ldap_receiver)
+			.fold(Ok((vec![], vec![], vec![])), |acc, entry_status| {
+				let (mut added, mut changed, mut removed) = acc?;
+				match entry_status {
+					EntryStatus::New(entry) => {
+						tracing::debug!("New entry: {:?}", entry);
+						added.push(self.parse_user(entry, None)?);
+					}
+					EntryStatus::Changed { old, new } => {
+						tracing::debug!("Changes found for {:?} -> {:?}", old, new);
+						changed.push((self.parse_user(old, None)?, self.parse_user(new, None)?));
+					}
+					EntryStatus::Removed(entry) => {
+						tracing::debug!("Deleted user {}", String::from_utf8_lossy(&entry));
+						removed.push(UserId::Login(String::from_utf8(entry.clone())?));
+					}
+				};
+				Ok((added, changed, removed))
+			})
+			.await
+	}
+
+	/// Construct a user from an LDAP SearchEntry
+	pub(crate) fn parse_user(&self, entry: SearchEntry, idp_id: Option<&str>) -> Result<User> {
+		let status_as_int = match read_search_entry(&entry, &self.ldap_config.attributes.status)? {
+			StringOrBytes::String(status) => status.parse::<i32>()?,
+			StringOrBytes::Bytes(status) => i32::from_be_bytes(
+				status.try_into().map_err(|_| anyhow!("failed to convert to i32 flag"))?,
```

Comment:
We are digging into old code, but anyway - this is not quite as straightforward as suggested, but this will do the trick:
```
i32::from_be_bytes(status.try_into().map_err(|err: Vec<u8>| {
	let err_string = String::from_utf8_lossy(&err).to_string();
	anyhow!(err_string).context("failed to convert to i32 flag")
})?)
```

Reply:
Sure, the discussion/investigation can be continued here: https://github.com/famedly/ldap-sync/issues/59